[LegalizeDAG] Optimize CodeGen for `ISD::CTLZ_ZERO_UNDEF`

Nirhar opened the pull request:
Previously we had the same instructions being generated for `ISD::CTLZ` and `ISD::CTLZ_ZERO_UNDEF` which did not take advantage of the fact that zero is an invalid input for `ISD::CTLZ_ZERO_UNDEF`. This commit separates codegen for the two cases to allow for the optimization for the latter case.

The details of the optimization are outlined in #82075

Fixes #82075

RKSimon commented on llvm/test/CodeGen/X86/clz.ll:
```diff
@@ -1517,13 +1510,11 @@ define i32 @cttz_i32_osize(i32 %x) optsize {
 ;
 ; X64-LABEL: cttz_i32_msize:
 ; X64:       # %bb.0:
-; X64-NOT:     rep
```
Add the NOT checks back (they were manually added after the update script is run)

Nirhar replied:
Done!

topperc commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -613,21 +613,42 @@ SDValue DAGTypeLegalizer::PromoteIntRes_CTLZ(SDNode *N) {
-                     ExtractLeadingBits, Mask, EVL);
+    SDValue Mask = N->getOperand(1);
+    SDValue EVL = N->getOperand(2);
+    Op = DAG.getNode(ISD::SHL, dl, NVT, Op, ShiftConst, Mask, EVL);
```
This needs to be ISD::VP_SHL. And it needs to be tested.

Nirhar replied:
@topperc I modified it to ISD::VP_SHL. However, how do I test it? Do I write a test that uses masked vector loads, and call the ctlz intrinsic on it?

topperc replied:
VP intrinsics are primarily used by RISC-V vectors. buildkite is reporting a RISC-V vector test failing.

  LLVM :: CodeGen/RISCV/rvv/ctlz-vp.ll

Nirhar replied:
Have updated the ctlz-vp.ll test using the update_tests script. Please check if they are correct

topperc commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -613,21 +613,42 @@ SDValue DAGTypeLegalizer::PromoteIntRes_CTLZ(SDNode *N) {
+                       ExtractLeadingBits, Mask, EVL);
+  } else if (CtlzOpcode == ISD::CTLZ_ZERO_UNDEF) {
+    // Zero Extend the argument
+    SDValue Op = ZExtPromotedInteger(N->getOperand(0));
```
The operand doesn't need to be zero extended does it? I think you're shifting out all the extended bits. So you can use GetPromotedInteger.

Nirhar replied:
Modified to `GetPromotedInteger`

topperc commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -613,21 +613,42 @@ SDValue DAGTypeLegalizer::PromoteIntRes_CTLZ(SDNode *N) {
+
+    // Op = Op << (sizeinbits(NVT) - sizeinbits(Old VT))
+    auto ShiftConst =
+        DAG.getConstant(NVT.getSizeInBits() - OVT.getSizeInBits(), dl, NVT);
```
This should use getShiftAmountConstant

Nirhar replied:
Modified to use `getShiftAmountConstant`

topperc commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -5003,6 +5002,22 @@ void SelectionDAGLegalize::PromoteNode(SDNode *Node) {
+    // different approach as compared to ISD::CTLZ
+
+    // Zero Extend the argument
+    Tmp1 = DAG.getNode(ISD::ZERO_EXTEND, dl, NVT, Node->getOperand(0));
```
This can be ANY_EXTEND

Nirhar replied:
Modified!

arsenm reviewed (commented):
Can you make the mirror change in the GlobalIsel lowering?

RKSimon commented:
@Nirhar reverse-ping

jayfoad commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -5003,6 +5002,22 @@ void SelectionDAGLegalize::PromoteNode(SDNode *Node) {
     }
     Results.push_back(DAG.getNode(ISD::TRUNCATE, dl, OVT, Tmp1));
     break;
+  case ISD::CTLZ_ZERO_UNDEF:
```
Remove the check for `Node->getOpcode() == ISD::CTLZ_ZERO_UNDEF` on line 4997 above?

Nirhar replied:
Done!

Nirhar commented:
@RKSimon, Sorry for the delay, I was under the weather last month. Will soon get back to working on this!

Nirhar commented on llvm/test/CodeGen/X86/pr38539.ll:
```diff
@@ -23,79 +23,84 @@ define void @f() nounwind {
 ; X86-NEXT:    pushl %esi
 ; X86-NEXT:    andl $-16, %esp
 ; X86-NEXT:    subl $176, %esp
-; X86-NEXT:    movzbl {{[0-9]+}}(%esp), %edx
```
@RKSimon @arsenm My PR somehow modifies the assembly generated in this test(which I think is not supposed to happen, given that this test has no calls to cltz intrinsic). Any heads up on why this could happen?

arsenm replied:
Some of the paths for divides introduce ctlzs

Nirhar commented:
> Can you make the mirror change in the GlobalIsel lowering?

Started to Look into it!

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -4982,15 +4981,30 @@ void SelectionDAGLegalize::PromoteNode(SDNode *Node) {
+    // Tmp1 = Tmp1 << (sizeinbits(NVT) - sizeinbits(Old VT))
+    Tmp1 = DAG.getNode(
+        ISD::SHL, dl, NVT, Tmp1,
+        DAG.getConstant(NVT.getSizeInBits() - OVT.getSizeInBits(), dl, NVT));
```
Use getShiftAmountConstant

Nirhar replied:
Done

topperc commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -642,21 +642,43 @@ SDValue DAGTypeLegalizer::PromoteIntRes_CTLZ(SDNode *N) {
-  // Zero extend to the promoted type and do the count there.
-  SDValue Op = ZExtPromotedInteger(N->getOperand(0));
+  unsigned CtlzOpcode = N->getOpcode();
+  if (CtlzOpcode == ISD::CTLZ) {
```
What about `ISD::VP_CTLZ`?

Nirhar replied:
Have accomodated it as well!

topperc commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -642,21 +642,43 @@ SDValue DAGTypeLegalizer::PromoteIntRes_CTLZ(SDNode *N) {
+    return DAG.getNode(ISD::VP_SUB, dl, NVT,
+                       DAG.getNode(N->getOpcode(), dl, NVT, Op, Mask, EVL),
+                       ExtractLeadingBits, Mask, EVL);
+  } else if (CtlzOpcode == ISD::CTLZ_ZERO_UNDEF) {
```
What about `ISD::VP_CTLZ_ZERO_UNDEF`?

Nirhar replied:
Have accomodated it as well!

topperc commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -642,21 +642,43 @@ SDValue DAGTypeLegalizer::PromoteIntRes_CTLZ(SDNode *N) {
+    return DAG.getNode(ISD::VP_SUB, dl, NVT,
+                       DAG.getNode(N->getOpcode(), dl, NVT, Op, Mask, EVL),
+                       ExtractLeadingBits, Mask, EVL);
+  } else if (CtlzOpcode == ISD::CTLZ_ZERO_UNDEF) {
```
No else after return https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return

Nirhar replied:
Hope the code formatting now is correct

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -642,21 +642,44 @@ SDValue DAGTypeLegalizer::PromoteIntRes_CTLZ(SDNode *N) {
+    // Subtract off the extra leading bits in the bigger type.
+    SDValue ExtractLeadingBits = DAG.getConstant(
+        NVT.getScalarSizeInBits() - OVT.getScalarSizeInBits(), dl, NVT);
+    if (!N->isVPOpcode())
+      return DAG.getNode(ISD::SUB, dl, NVT,
+                         DAG.getNode(N->getOpcode(), dl, NVT, Op),
+                         ExtractLeadingBits);
+    SDValue Mask = N->getOperand(1);
+    SDValue EVL = N->getOperand(2);
+    return DAG.getNode(ISD::VP_SUB, dl, NVT,
+                       DAG.getNode(N->getOpcode(), dl, NVT, Op, Mask, EVL),
+                       ExtractLeadingBits, Mask, EVL);
```
We really need a better API for emit-VP-or-not-VP

arsenm commented on llvm/test/CodeGen/AArch64/ctlz_zero_undef.ll:
```diff
@@ -0,0 +1,18 @@
+  %0 = call i8 @llvm.ctlz.i8(i8 %self, i1 true)
+  %_0 = zext i8 %0 to i32
+  ret i32 %_0
+}
```
Missing newline at end of file

Nirhar replied:
Added!

arsenm commented on llvm/test/CodeGen/AArch64/ctlz_zero_undef.ll:
```diff
@@ -0,0 +1,18 @@
+; CHECK-NEXT:    clz w0, w8
+; CHECK-NEXT:    ret
+start:
+  %self = load i8, ptr %num, align 1
```
Simpler to have an i8 function argument than the load

Nirhar replied:
Yes, I agree, modified it!

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -4982,15 +4981,30 @@ void SelectionDAGLegalize::PromoteNode(SDNode *Node) {
+
+    // Perform the larger operation
+    Tmp1 = DAG.getNode(Node->getOpcode(), dl, NVT, Tmp1);
+    Results.push_back(DAG.getNode(ISD::TRUNCATE, dl, OVT, Tmp1));
```
I think I never got around to fixing this in the DAG, but all of these bitcount operations in principle could have a different return type from the source operand. It would be a bit better to truncate to the result type, and use the source type as the old type throughout

Nirhar replied:
I see that we truncate to OVT bitsize almost everywhere in PromoteNode. Hence I feel your suggestion here is better as a follow up patch

arsenm reviewed (commented):
Does GlobalIsel need the equivalent handling?

Nirhar commented on llvm/test/CodeGen/AArch64/ctlz_zero_undef.ll:
```diff
@@ -0,0 +1,18 @@
+; CHECK-NEXT:    ret
+start:
+  %self = load i8, ptr %num, align 1
+  %0 = call i8 @llvm.ctlz.i8(i8 %self, i1 true)
```
Added some new tests as you suggested, please check!

Nirhar commented:
Modified GlobalISel code as well to include the new optimization

RKSimon commented:
@Nirhar reverse-ping?

Commit:
[LegalizeDAG] Optimize CodeGen for `ISD::CTLZ_ZERO_UNDEF`

Previously we had the same instructions being generated for
`ISD::CTLZ` and `ISD::CTLZ_ZERO_UNDEF` which did not take
advantage of the fact that zero is an invalid input for
`ISD::CTLZ_ZERO_UNDEF`. This commit separates codegen for
the two cases to allow for the optimization for the latter case.

The details of the optimization are outlined in #82075

Fixes #82075

Nirhar commented on llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:
```diff
@@ -2458,13 +2458,20 @@ LegalizerHelper::widenScalar(MachineInstr &MI, unsigned TypeIdx, LLT WideTy) {
+
+    if (MI.getOpcode() == TargetOpcode::G_CTLZ_ZERO_UNDEF) {
+      // An optimization where the result is the CTLZ after the left shift by
+      // (Difference in widety and current ty).
```
Added!

Nirhar commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -5047,15 +5046,30 @@ void SelectionDAGLegalize::PromoteNode(SDNode *Node) {
+
+    // Tmp1 = Tmp1 << (sizeinbits(NVT) - sizeinbits(Old VT))
+    Tmp2 = DAG.getShiftAmountConstant(NVT.getSizeInBits() - OVT.getSizeInBits(),
+                                      Tmp1.getValueType(), dl);
```
Modified!

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -5115,6 +5114,22 @@ void SelectionDAGLegalize::PromoteNode(SDNode *Node) {
+    // different approach as compared to ISD::CTLZ
+
+    // Any Extend the argument
+    Tmp1 = DAG.getNode(ISD::ANY_EXTEND, dl, NVT, Node->getOperand(0));
```
Nicer to use locally defined, better named temporary variables

Nirhar replied:
I agree, but I saw that Tmp1, Tmp2 etc are explicitly used for temporary nodes in other cases in this switch statement, which I guess comes from the limitation that we aren't allowed to declare new variables inside a case. Hence I decided to follow the same convention.

jayfoad replied:
> we aren't allowed to declare new variables inside a case

Yes you are, just put braces around the whole case.

Nirhar replied:
Aah didn't know about this fact, thanks! Have updated teh patch with new variable names!

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -655,24 +655,46 @@ SDValue DAGTypeLegalizer::PromoteIntRes_CTLZ(SDNode *N) {
-      NVT.getScalarSizeInBits() - OVT.getScalarSizeInBits(), dl, NVT);
-  if (!N->isVPOpcode()) {
+  unsigned CtlzOpcode = N->getOpcode();
+  if (CtlzOpcode == ISD::CTLZ || CtlzOpcode == ISD::VP_CTLZ) {
```
If we properly decoupled the source and result types for these operations, like in globalisel, you would move this to the operand promote.

It's really unfortunate you need to repeat all of this logic in PromoteIntRes

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -655,24 +655,46 @@ SDValue DAGTypeLegalizer::PromoteIntRes_CTLZ(SDNode *N) {
 
-  SDValue Mask = N->getOperand(1);
-  SDValue EVL = N->getOperand(2);
-  // Zero extend to the promoted type and do the count there.
-  SDValue Op = VPZExtPromotedInteger(N->getOperand(0), Mask, EVL);
-  return DAG.getNode(ISD::VP_SUB, dl, NVT,
-                     DAG.getNode(N->getOpcode(), dl, NVT, Op, Mask, EVL),
-                     ExtractLeadingBits, Mask, EVL);
+    SDValue Mask = N->getOperand(1);
+    SDValue EVL = N->getOperand(2);
+    Op = DAG.getNode(ISD::VP_SHL, dl, NVT, Op, ShiftConst, Mask, EVL);
+    return DAG.getNode(CtlzOpcode, dl, NVT, Op, Mask, EVL);
```
We really need a more manageable API for dealing with the VP_ path at the same time

arsenm reviewed (commented):
LGTM with nit (modulo the DAG replication, which I don't have a suggestion to avoid)

Nirhar commented:
I do not understand why the CI failed here: https://buildkite.com/llvm-project/github-pull-requests/builds/76944#019068e3-bd93-4ad9-ba55-817fdabbe0eb. Looks like all the tests passed

jayfoad commented:
> I do not understand why the CI failed here: https://buildkite.com/llvm-project/github-pull-requests/builds/76944#019068e3-bd93-4ad9-ba55-817fdabbe0eb. Looks like all the tests passed

Further up in the build log there is:
```
Failed Tests (1):
  BOLT :: X86/reader-stale-yaml-std.test
```
and:
```
Timed Out Tests (1):
  lldb-api :: functionalities/fork/concurrent_vfork/TestConcurrentVFork.py
```

RKSimon commented:
The lldb timeout is an intermittent fail - nothing to do with your patch.

Not sure about the bolt regression, but I think this is safe to commit - the bolt team are pretty good at fixing anything that comes up due to their project dependencies (you're not expected to fix everything yourself).
